### PR TITLE
Replaced mutation observer with root element zone changes

### PR DESCRIPTION
**Pull request by sumitarora**
Should fix: https://github.com/rangle/batarangle/issues/181 and https://github.com/rangle/batarangle/issues/164

### CODE REVIEW

File: src/backend/adapters/base.ts [hunk 1]
```diff
@@ -117,5 +117,5 @@ export abstract class BaseAdapter {
 
   abstract serializeComponent(el: any, event: string): TreeNode;
 
-  abstract cleanup(): void;
+  // abstract cleanup(): void;
```
Comment (igor-ka) [inline src/backend/adapters/base.ts]: let's stay away from keeping commented out code, we have history with source control and can always go back if needed. I would remove it.

Comment (sumitarora) [inline src/backend/adapters/base.ts]: :+1:

File: src/backend/backend.ts [hunk 2]
```diff
@@ -30,13 +30,6 @@ window.addEventListener('message', function(event) {
 
     if (event.data.message.message.actionType ===
```
Comment (igor-ka) [inline src/backend/backend.ts]: the message.message is not a part of this PR, but this is a leftover from the days when I used to contribute to this :), we should unnest this sometime in the future.

File: src/backend/backend.ts [hunk 3]
```diff
@@ -50,12 +43,16 @@ window.addEventListener('message', function(event) {
       const highlightStr = '[batarangle-id=\"' +
         event.data.message.message.node.id + '\"]';
 
-      (<HTMLElement>document.querySelector(highlightStr)).scrollIntoView();
+      const element: any = <HTMLElement>document.querySelector(highlightStr);
```
Comment (igor-ka) [inline src/backend/backend.ts]: Do we need a cast to HTML element here? does querySelector return any or something else?

File: src/backend/controllers/dom.ts [hunk 4]
```diff
@@ -69,7 +83,11 @@ export class DomController extends BaseController {
     }
 
     if (ch) {
-      ch.sendMessage({ type: 'model_change', payload: this.model});
+      this.callToRenderTree.next({
+        channel: ch,
+        message: { type: 'model_change', payload: this.model }
+      });
+      // ch.sendMessage({ type: 'model_change', payload: this.model});
```
Comment (igor-ka) [inline src/backend/controllers/dom.ts]: I would remove the commented out code here.

File: src/backend/controllers/dom.ts [hunk 5]
```diff
@@ -41,6 +44,17 @@ export class DomController extends BaseController {
     this.model = [];
     this.adapter = adapter;
     this.channel = channel;
+
+    this.callToRenderTree = new Rx.Subject();
+    this.callToRenderTree
+      .debounce((x) => {
+        return Rx.Observable.timer(250);
```
Comment (igor-ka) [inline src/backend/controllers/dom.ts]: I would remove the curlies and the round brackets on the argument to keep it as a one liner

File: src/backend/controllers/dom.ts [hunk 6]
```diff
@@ -19,6 +19,8 @@ import { AdapterEvent } from '../adapters/base';
 import { Angular2Adapter } from '../adapters/angular2';
 import { BaseController } from './base';
 
+import * as Rx from 'rxjs';
```
Comment (igor-ka) [inline src/backend/controllers/dom.ts]: I would import only what is needed from rxjs as opposed to the whole thing

File: src/backend/adapters/angular2.ts [hunk 7]
```diff
@@ -32,22 +32,48 @@ import { DirectiveProvider } from 'angular2/src/core/linker/element';
...
       '0',
       this._emitNativeElement);
 
-    this._trackChanges(root);
+    this._trackAngularChanges(ng.probe(root));
+  }
+
+  _trackAngularChanges(rootNgProbe: any) {
+    const ngZone = rootNgProbe.inject(ng.coreTokens.NgZone);
+    ngZone.onEventDone.subscribe(() => { this._onEventDone.next();});
```
Comment (igor-ka) [inline src/backend/adapters/angular2.ts]: curlies are not needed here and I would remove them

File: src/backend/adapters/angular2.ts [hunk 8]
```diff
@@ -32,22 +32,48 @@ import { DirectiveProvider } from 'angular2/src/core/linker/element';
...
-
   _tree: any = {};
+  _onEventDone: any;
+
+  constructor() {
+    super();
+    this._onEventDone = new Rx.Subject();
+
+    this._onEventDone
+      .debounce((x) => {
```
Comment (igor-ka) [inline src/backend/adapters/angular2.ts]: I would make this a one liner as it is very simple

File: src/backend/adapters/angular2.ts [hunk 9]
```diff
@@ -32,22 +32,48 @@ import { DirectiveProvider } from 'angular2/src/core/linker/element';
 import { Description } from '../utils/description';
 
 import {DirectiveResolver} from '../directive-resolver';
+import * as Rx from 'rxjs';
```
Comment (igor-ka) [inline src/backend/adapters/angular2.ts]: I would import only what is needed from rxjs here

File: src/backend/adapters/angular2.ts [hunk 10]
```diff
@@ -24,7 +24,7 @@
  * Supports up to 2.0.0-beta-1
  */
 
-declare var ng: { probe: Function };
+declare var ng: { probe: Function, coreTokens: any };
```
Comment (igor-ka) [inline src/backend/adapters/angular2.ts]: I am wondering if the typings for those are defined somewhere within angular's generated *.d.ts files? This way we could get rid of the ambient declaration here, but this is just a nice to have :)

**Comment by igor-ka** [comment]
@sumitarora overall LGTM minus the comments. Otherwise I am not sure if I followed what is happening here 100% as I am a but out of the loop with regards to what API they provided us with regards to being notified of changes to the components.

I am giving it a :+1: , but let's have @vanessayuenn or @bertrandk give a second look through.
Otherwise this is great, good job!

File: src/backend/adapters/angular2.ts (hunk 10 above)
Comment (bertrandk) [inline src/backend/adapters/angular2.ts]: Good call...I need to double check but for some reason I seem to recall that it was typed but the typing weren't exported. Probably has changed since this is a public API.

File: src/backend/backend.ts (hunk 3 above)
Comment (bertrandk) [inline src/backend/backend.ts]: I believe it returns Element instead of HTMLElement.

**Comment by bertrandk** [comment]
The suspicion is that we're leaking a lot of DOM nodes in this routine. @sumitarora, did you plan on addressing that in this PR or treating it separately?

In any case, aside from the small cleanup advocated, this looks good to me.

File: src/backend/backend.ts (hunk 2 above)
Comment (sumitarora) [inline src/backend/backend.ts]: will do in this issue https://github.com/rangle/batarangle/issues/142

File: src/backend/controllers/dom.ts (hunk 6 above)
Comment (sumitarora) [inline src/backend/controllers/dom.ts]: using it in creating debounce observable

File: src/backend/adapters/angular2.ts (hunk 9 above)
Comment (sumitarora) [inline src/backend/adapters/angular2.ts]: using it in creating debounce observable and subject

**Comment by sumitarora** [comment]
@bertrandk I will do another PR for DOM issues

File: src/backend/adapters/angular2.ts [hunk 11]
```diff
@@ -109,11 +133,6 @@ export class Angular2Adapter extends BaseAdapter {
     };
   }
 
-  cleanup(): void {
```
Comment (vanessayuenn) [inline src/backend/adapters/angular2.ts]: the new listener doesn't need to be disconnected/unsubscribed?

**Comment by vanessayuenn** [comment]
only had one question about the unsubscribe, but it looks good otherwise. :+1:

File: src/backend/controllers/dom.ts (hunk 6 above)
Comment (igor-ka) [inline src/backend/controllers/dom.ts]: You can import `debounce` operators on its own as well.

File: src/backend/adapters/angular2.ts (hunk 9 above)
Comment (igor-ka) [inline src/backend/adapters/angular2.ts]: You can import `debounce` operators on its own as well.

Comment (sumitarora) [inline src/backend/adapters/angular2.ts]: will do in another PR

File: src/backend/controllers/dom.ts (hunk 6 above)
Comment (sumitarora) [inline src/backend/controllers/dom.ts]: will do in another PR

File: src/backend/adapters/angular2.ts (hunk 9 above)
Comment (igor-ka) [inline src/backend/adapters/angular2.ts]: :+1:

File: src/backend/controllers/dom.ts (hunk 6 above)
Comment (igor-ka) [inline src/backend/controllers/dom.ts]: :+1:
